File: Legacy/Source/Modulo_InventANDobj.vb
Option Strict Off
Option Explicit On
Module InvNpc
    'Modulo Inv & Obj
    'Modulo para controlar los objetos y los inventarios.
    Public Function TirarItemAlPiso(ByRef Pos As WorldPos, ByRef Obj As Obj, Optional ByRef NotPirata As Boolean = True) _
        As WorldPos
        '***************************************************
        'Author: Unknown
        'Last Modification: -
        '
        '***************************************************

        Try

            Dim NuevaPos As WorldPos
            NuevaPos.X = 0
            NuevaPos.Y = 0

            Tilelibre(Pos, NuevaPos, Obj, NotPirata, True)
            If NuevaPos.X <> 0 And NuevaPos.Y <> 0 Then
                Call MakeObj(Obj, Pos.Map, NuevaPos.X, NuevaPos.Y)
            End If
            'UPGRADE_WARNING: No se puede resolver la propiedad predeterminada del objeto TirarItemAlPiso. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"'
            TirarItemAlPiso = NuevaPos


        Catch ex As Exception
            Console.WriteLine("Error in TirarItemAlPiso: " & ex.Message)

        End Try
    End Function

    Public Sub NPC_TIRAR_ITEMS(ByRef npc As npc, IsPretoriano As Boolean)
        '***************************************************
        'Autor: Unknown (orginal version)
        'Last Modification: 28/11/2009
        'Give away npc's items.
        '28/11/2009: ZaMa - Implementado drops complejos
        '02/04/2010: ZaMa - Los pretos vuelven a tirar oro.
        '***************************************************
        Try

            Dim i As Byte
            Dim MiObj As Obj
            Dim NroDrop As Short
            Dim Random As Short
            Dim ObjIndex As Short
            With npc


                ' Tira todo el inventario
                If IsPretoriano Then
                    For i = 1 To MAX_INVENTORY_SLOTS
                        If .Invent.Object_Renamed(i).ObjIndex > 0 Then
                            MiObj.Amount = .Invent.Object_Renamed(i).Amount
                            MiObj.ObjIndex = .Invent.Object_Renamed(i).ObjIndex
                            Call TirarItemAlPiso(.Pos, MiObj)
                        End If
                    Next i

                    ' Dropea oro?
                    If .GiveGLD > 0 Then Call TirarOroNpc(.GiveGLD, .Pos)

                    Exit Sub
                End If

                Random = RandomNumber(1, 100)

                ' Tiene 10% de prob de no tirar nada
                If Random <= 90 Then
                    NroDrop = 1

                    If Random <= 10 Then
                        NroDrop = NroDrop + 1

                        For i = 1 To 3
                            ' 10% de ir pasando de etapas
                            If RandomNumber(1, 100) <= 10 Then
                                NroDrop = NroDrop + 1
                            Else
                                Exit For
                            End If
                        Next i

                    End If


                    ObjIndex = .Drop(NroDrop).ObjIndex
                    If ObjIndex > 0 Then

                        If ObjIndex = iORO Then
                            Call TirarOroNpc(.Drop(NroDrop).Amount, npc.Pos)
                        Else
                            MiObj.Amount = .Drop(NroDrop).Amount
                            MiObj.ObjIndex = .Drop(NroDrop).ObjIndex

                            Call TirarItemAlPiso(.Pos, MiObj)
                        End If
                    End If

                End If

            End With

        Catch ex As Exception
            Console.WriteLine("Error in TirarItemAlPiso: " & ex.Message)
        End Try
    End Sub

    Function QuedanItems(NpcIndex As Short, ObjIndex As Short) As Boolean
        '***************************************************
        'Author: Unknown
        'Last Modification: -
        '
        '***************************************************

        Try

            Dim i As Short
            If Npclist(NpcIndex).Invent.NroItems > 0 Then
                For i = 1 To MAX_INVENTORY_SLOTS
                    If Npclist(NpcIndex).Invent.Object_Renamed(i).ObjIndex = ObjIndex Then
                        QuedanItems = True
                        Exit Function
                    End If
                Next
            End If
            QuedanItems = False

        Catch ex As Exception
            Console.WriteLine("Error in QuedanItems: " & ex.Message)
        End Try
    End Function

    ''
    ' Gets the amount of a certain item that an npc has.
    '
    ' @param npcIndex Specifies reference to npcmerchant
    ' @param ObjIndex Specifies reference to object
    ' @return   The amount of the item that the npc has
    ' @remarks This function reads the Npc.dat file
    Function EncontrarCant(NpcIndex As Short, ObjIndex As Short) As Short
        '***************************************************
        'Author: Unknown
        'Last Modification: 03/09/08
        'Last Modification By: Marco Vanotti (Marco)
        ' - 03/09/08 EncontrarCant now returns 0 if the npc doesn't have it (Marco)
        '***************************************************
        Try
            'Devuelve la cantidad original del obj de un npc

            Dim ln, npcfile As String
            Dim i As Short

            npcfile = DatPath & "NPCs.dat"

            For i = 1 To MAX_INVENTORY_SLOTS
                ln = GetVar(npcfile, "NPC" & Npclist(NpcIndex).Numero, "Obj" & i)
                If ObjIndex = Val(ReadField(1, ln, 45)) Then
                    EncontrarCant = Val(ReadField(2, ln, 45))
                    Exit Function
                End If
            Next

            EncontrarCant = 0

        Catch ex As Exception
            Console.WriteLine("Error in reads: " & ex.Message)
        End Try
    End Function

    Sub ResetNpcInv(NpcIndex As Short)
        '***************************************************
        'Author: Unknown
        'Last Modification: -
        '
        '***************************************************

        Try

            Dim i As Short

            With Npclist(NpcIndex)
                .Invent.NroItems = 0

                For i = 1 To MAX_INVENTORY_SLOTS
                    .Invent.Object_Renamed(i).ObjIndex = 0
                    .Invent.Object_Renamed(i).Amount = 0
                Next i

                .InvReSpawn = 0
            End With

        Catch ex As Exception
            Console.WriteLine("Error in ResetNpcInv: " & ex.Message)
        End Try
    End Sub

    ''
    ' Removes a certain amount of items from a slot of an npc's inventory
    '
    ' @param npcIndex Specifies reference to npcmerchant
    ' @param Slot Specifies reference to npc's inventory's slot
    ' @param antidad Specifies amount of items that will be removed
    Sub QuitarNpcInvItem(NpcIndex As Short, Slot As Byte, Cantidad As Short)
        '***************************************************
        'Author: Unknown
        'Last Modification: 23/11/2009
        'Last Modification By: Marco Vanotti (Marco)
        ' - 03/09/08 Now this sub checks that te npc has an item before respawning it (Marco)
        '23/11/2009: ZaMa - Optimizacion de codigo.
        '***************************************************
        Dim ObjIndex As Short
        Dim iCant As Short

        With Npclist(NpcIndex)
            ObjIndex = .Invent.Object_Renamed(Slot).ObjIndex

            'Quita un Obj
            If ObjData_Renamed(.Invent.Object_Renamed(Slot).ObjIndex).Crucial = 0 Then
                .Invent.Object_Renamed(Slot).Amount = .Invent.Object_Renamed(Slot).Amount - Cantidad

                If .Invent.Object_Renamed(Slot).Amount <= 0 Then
                    .Invent.NroItems = .Invent.NroItems - 1
                    .Invent.Object_Renamed(Slot).ObjIndex = 0
                    .Invent.Object_Renamed(Slot).Amount = 0
                    If .Invent.NroItems = 0 And .InvReSpawn <> 1 Then
                        Call CargarInvent(NpcIndex) 'Reponemos el inventario
                    End If
                End If
            Else
                .Invent.Object_Renamed(Slot).Amount = .Invent.Object_Renamed(Slot).Amount - Cantidad

                If .Invent.Object_Renamed(Slot).Amount <= 0 Then
                    .Invent.NroItems = .Invent.NroItems - 1
                    .Invent.Object_Renamed(Slot).ObjIndex = 0
                    .Invent.Object_Renamed(Slot).Amount = 0

                    If Not QuedanItems(NpcIndex, ObjIndex) Then
                        'Check if the item is in the npc's dat.
                        iCant = EncontrarCant(NpcIndex, ObjIndex)
                        If iCant Then
                            .Invent.Object_Renamed(Slot).ObjIndex = ObjIndex
                            .Invent.Object_Renamed(Slot).Amount = iCant
                            .Invent.NroItems = .Invent.NroItems + 1
                        End If
                    End If

                    If .Invent.NroItems = 0 And .InvReSpawn <> 1 Then
                        Call CargarInvent(NpcIndex) 'Reponemos el inventario
                    End If
                End If
            End If
        End With
    End Sub

    Sub CargarInvent(NpcIndex As Short)
        '***************************************************
        'Author: Unknown
        'Last Modification: -
        '
        '***************************************************

        'Vuelve a cargar el inventario del npc NpcIndex
        Dim LoopC As Short
        Dim ln As String
        Dim npcfile As String

        npcfile = DatPath & "NPCs.dat"

        With Npclist(NpcIndex)
            .Invent.NroItems = Val(GetVar(npcfile, "NPC" & .Numero, "NROITEMS"))

            For LoopC = 1 To .Invent.NroItems
                ln = GetVar(npcfile, "NPC" & .Numero, "Obj" & LoopC)
                .Invent.Object_Renamed(LoopC).ObjIndex = Val(ReadField(1, ln, 45))
                .Invent.Object_Renamed(LoopC).Amount = Val(ReadField(2, ln, 45))

            Next LoopC
        End With
    End Sub


    Public Sub TirarOroNpc(Cantidad As Integer, ByRef Pos As WorldPos)
        '***************************************************
        'Autor: ZaMa
        'Last Modification: 13/02/2010
        '***************************************************
        Try

            Dim i As Byte
            Dim MiObj As Obj
            Dim RemainingGold As Integer
            If Cantidad > 0 Then

                RemainingGold = Cantidad

                While (RemainingGold > 0)

                    ' Tira pilon de 10k
                    If RemainingGold > MAX_INVENTORY_OBJS Then
                        MiObj.Amount = MAX_INVENTORY_OBJS
                        RemainingGold = RemainingGold - MAX_INVENTORY_OBJS

                        ' Tira lo que quede
                    Else
                        MiObj.Amount = RemainingGold
                        RemainingGold = 0
                    End If

                    MiObj.ObjIndex = iORO

                    Call TirarItemAlPiso(Pos, MiObj)
                End While
            End If


        Catch ex As Exception
            Console.WriteLine("Error in NPC_TIRAR_ITEMS: " & ex.Message)
            Call LogError("Error en TirarOro. Error " & Err.Number & " : " & Err.Description)
        End Try
    End Sub
End Module

File: Legacy/Source/Modulo_InventANDobj.cs
using System;
using System.Runtime.InteropServices;

namespace Legacy
{
    static class InvNpc
    {
        // Modulo Inv & Obj
        // Modulo para controlar los objetos y los inventarios.
        public static Declaraciones.WorldPos TirarItemAlPiso(ref Declaraciones.WorldPos Pos, ref Declaraciones.Obj Obj, [Optional, DefaultParameterValue(true)] ref bool NotPirata)
        {
            Declaraciones.WorldPos TirarItemAlPisoRet = default;
            // ***************************************************
            // Author: Unknown
            // Last Modification: -
            // 
            // ***************************************************

            try
            {

                var NuevaPos = default(Declaraciones.WorldPos);
                NuevaPos.X = 0;
                NuevaPos.Y = 0;

                bool argTierra = true;
                UsUaRiOs.Tilelibre(ref Pos, ref NuevaPos, ref Obj, ref NotPirata, ref argTierra);
                if (NuevaPos.X != 0 & NuevaPos.Y != 0)
                {
                    InvUsuario.MakeObj(ref Obj, Pos.Map, NuevaPos.X, NuevaPos.Y);
                }
                // UPGRADE_WARNING: No se puede resolver la propiedad predeterminada del objeto TirarItemAlPiso. Haga clic aquí para obtener más información: 'ms-help://MS.VSCC.v90/dv_commoner/local/redirect.htm?keyword="6A50421D-15FE-4896-8A1B-2EC21E9037B2"'
                TirarItemAlPisoRet = NuevaPos;
            }


            catch (Exception ex)
            {
                Console.WriteLine("Error in TirarItemAlPiso: " + ex.Message);

            }

            return TirarItemAlPisoRet;
        }

        public static void NPC_TIRAR_ITEMS(ref Declaraciones.npc npc, bool IsPretoriano)
        {
            // ***************************************************
            // Autor: Unknown (orginal version)
            // Last Modification: 28/11/2009
            // Give away npc's items.
            // 28/11/2009: ZaMa - Implementado drops complejos
            // 02/04/2010: ZaMa - Los pretos vuelven a tirar oro.
            // ***************************************************
            try
            {

                byte i;
                Declaraciones.Obj MiObj;
                short NroDrop;
                short Random;
                short ObjIndex;


                // Tira todo el inventario
                if (IsPretoriano)
                {
                    for (i = 1; i <= Declaraciones.MAX_INVENTORY_SLOTS; i++)
                    {
                        if (npc.Invent.Object_Renamed[i].ObjIndex > 0)
                        {
                            MiObj.Amount = npc.Invent.Object_Renamed[i].Amount;
                            MiObj.ObjIndex = npc.Invent.Object_Renamed[i].ObjIndex;
                            bool argNotPirata = true;
                            TirarItemAlPiso(ref npc.Pos, ref MiObj, NotPirata: ref argNotPirata);
                        }
                    }

                    // Dropea oro?
                    if (npc.GiveGLD > 0)
                        TirarOroNpc(npc.GiveGLD, ref npc.Pos);

                    return;
                }

                Random = Convert.ToInt16(Matematicas.RandomNumber(1, 100));

                // Tiene 10% de prob de no tirar nada
                if (Random <= 90)
                {
                    NroDrop = 1;

                    if (Random <= 10)
                    {
                        NroDrop = Convert.ToInt16(NroDrop + 1);

                        for (i = 1; i <= 3; i++)
                        {
                            // 10% de ir pasando de etapas
                            if (Matematicas.RandomNumber(1, 100) <= 10)
                            {
                                NroDrop = Convert.ToInt16(NroDrop + 1);
                            }
                            else
                            {
                                break;
                            }
                        }

                    }


                    ObjIndex = npc.Drop[NroDrop].ObjIndex;
                    if (ObjIndex > 0)
                    {

                        if (ObjIndex == Declaraciones.iORO)
                        {
                            TirarOroNpc(npc.Drop[NroDrop].Amount, ref npc.Pos);
                        }
                        else
                        {
                            MiObj.Amount = Convert.ToInt16(npc.Drop[NroDrop].Amount);
                            MiObj.ObjIndex = npc.Drop[NroDrop].ObjIndex;

                            bool argNotPirata1 = true;
                            TirarItemAlPiso(ref npc.Pos, ref MiObj, NotPirata: ref argNotPirata1);
                        }
                    }


                }
            }

            catch (Exception ex)
            {
                Console.WriteLine("Error in TirarItemAlPiso: " + ex.Message);
            }
        }

        public static bool QuedanItems(short NpcIndex, short ObjIndex)
        {
            bool QuedanItemsRet = default;
            // ***************************************************
            // Author: Unknown
            // Last Modification: -
            // 
            // ***************************************************

            try
            {

                short i;
                if (Declaraciones.Npclist[NpcIndex].Invent.NroItems > 0)
                {
                    for (i = 1; i <= Declaraciones.MAX_INVENTORY_SLOTS; i++)
                    {
                        if (Declaraciones.Npclist[NpcIndex].Invent.Object_Renamed[i].ObjIndex == ObjIndex)
                        {
                            QuedanItemsRet = true;
                            return QuedanItemsRet;
                        }
                    }
                }
                QuedanItemsRet = false;
            }

            catch (Exception ex)
            {
                Console.WriteLine("Error in QuedanItems: " + ex.Message);
            }

            return QuedanItemsRet;
        }

        // '
        // Gets the amount of a certain item that an npc has.
        // 
        // @param npcIndex Specifies reference to npcmerchant
        // @param ObjIndex Specifies reference to object
        // @return   The amount of the item that the npc has
        // @remarks This function reads the Npc.dat file
        public static short EncontrarCant(short NpcIndex, short ObjIndex)
        {
            short EncontrarCantRet = default;
            // ***************************************************
            // Author: Unknown
            // Last Modification: 03/09/08
            // Last Modification By: Marco Vanotti (Marco)
            // - 03/09/08 EncontrarCant now returns 0 if the npc doesn't have it (Marco)
            // ***************************************************
            try
            {
                // Devuelve la cantidad original del obj de un npc

                string ln, npcfile;
                short i;

                npcfile = Declaraciones.DatPath + "NPCs.dat";

                for (i = 1; i <= Declaraciones.MAX_INVENTORY_SLOTS; i++)
                {
                    int argEmptySpaces = 1024;
                    ln = ES.GetVar(npcfile, "NPC" + Declaraciones.Npclist[NpcIndex].Numero, "Obj" + i, EmptySpaces: ref argEmptySpaces);
                    if (ObjIndex == Convert.ToInt16(Migration.ParseVal(General.ReadField(1, ref ln, 45))))
                    {
                        EncontrarCantRet = Convert.ToInt16(Migration.ParseVal(General.ReadField(2, ref ln, 45)));
                        return EncontrarCantRet;
                    }
                }

                EncontrarCantRet = 0;
            }

            catch (Exception ex)
            {
                Console.WriteLine("Error in reads: " + ex.Message);
            }

            return EncontrarCantRet;
        }

        public static void ResetNpcInv(short NpcIndex)
        {
            // ***************************************************
            // Author: Unknown
            // Last Modification: -
            // 
            // ***************************************************

            try
            {

                short i;

                {
                    ref var withBlock = ref Declaraciones.Npclist[NpcIndex];
                    withBlock.Invent.NroItems = 0;

                    for (i = 1; i <= Declaraciones.MAX_INVENTORY_SLOTS; i++)
                    {
                        withBlock.Invent.Object_Renamed[i].ObjIndex = 0;
                        withBlock.Invent.Object_Renamed[i].Amount = 0;
                    }

                    withBlock.InvReSpawn = 0;
                }
            }

            catch (Exception ex)
            {
                Console.WriteLine("Error in ResetNpcInv: " + ex.Message);
            }
        }

        // '
        // Removes a certain amount of items from a slot of an npc's inventory
        // 
        // @param npcIndex Specifies reference to npcmerchant
        // @param Slot Specifies reference to npc's inventory's slot
        // @param antidad Specifies amount of items that will be removed
        public static void QuitarNpcInvItem(short NpcIndex, byte Slot, short Cantidad)
        {
            // ***************************************************
            // Author: Unknown
            // Last Modification: 23/11/2009
            // Last Modification By: Marco Vanotti (Marco)
            // - 03/09/08 Now this sub checks that te npc has an item before respawning it (Marco)
            // 23/11/2009: ZaMa - Optimizacion de codigo.
            // ***************************************************
            short ObjIndex;
            short iCant;

            {
                ref var withBlock = ref Declaraciones.Npclist[NpcIndex];
                ObjIndex = withBlock.Invent.Object_Renamed[Slot].ObjIndex;

                // Quita un Obj
                if (Declaraciones.ObjData_Renamed[withBlock.Invent.Object_Renamed[Slot].ObjIndex].Crucial == 0)
                {
                    withBlock.Invent.Object_Renamed[Slot].Amount = Convert.ToInt16((short)(withBlock.Invent.Object_Renamed[Slot].Amount - Cantidad));

                    if (withBlock.Invent.Object_Renamed[Slot].Amount <= 0)
                    {
                        withBlock.Invent.NroItems = Convert.ToInt16(withBlock.Invent.NroItems - 1);
                        withBlock.Invent.Object_Renamed[Slot].ObjIndex = 0;
                        withBlock.Invent.Object_Renamed[Slot].Amount = 0;
                        if (withBlock.Invent.NroItems == 0 & withBlock.InvReSpawn != 1)
                        {
                            CargarInvent(NpcIndex); // Reponemos el inventario
                        }
                    }
                }
                else
                {
                    withBlock.Invent.Object_Renamed[Slot].Amount = Convert.ToInt16((short)(withBlock.Invent.Object_Renamed[Slot].Amount - Cantidad));

                    if (withBlock.Invent.Object_Renamed[Slot].Amount <= 0)
                    {
                        withBlock.Invent.NroItems = Convert.ToInt16(withBlock.Invent.NroItems - 1);
                        withBlock.Invent.Object_Renamed[Slot].ObjIndex = 0;
                        withBlock.Invent.Object_Renamed[Slot].Amount = 0;

                        if (!QuedanItems(NpcIndex, ObjIndex))
                        {
                            // Check if the item is in the npc's dat.
                            iCant = Convert.ToInt16(EncontrarCant(NpcIndex, ObjIndex));
                            if (iCant != 0)
                            {
                                withBlock.Invent.Object_Renamed[Slot].ObjIndex = ObjIndex;
                                withBlock.Invent.Object_Renamed[Slot].Amount = iCant;
                                withBlock.Invent.NroItems = Convert.ToInt16(withBlock.Invent.NroItems + 1);
                            }
                        }

                        if (withBlock.Invent.NroItems == 0 & withBlock.InvReSpawn != 1)
                        {
                            CargarInvent(NpcIndex); // Reponemos el inventario
                        }
                    }
                }
            }
        }

        public static void CargarInvent(short NpcIndex)
        {
            // ***************************************************
            // Author: Unknown
            // Last Modification: -
            // 
            // ***************************************************

            // Vuelve a cargar el inventario del npc NpcIndex
            short LoopC;
            string ln;
            string npcfile;

            npcfile = Declaraciones.DatPath + "NPCs.dat";

            {
                ref var withBlock = ref Declaraciones.Npclist[NpcIndex];
                int argEmptySpaces = 1024;
                withBlock.Invent.NroItems = Convert.ToInt16(Migration.ParseVal(ES.GetVar(npcfile, "NPC" + withBlock.Numero, "NROITEMS", EmptySpaces: ref argEmptySpaces)));

                var loopTo = withBlock.Invent.NroItems;
                for (LoopC = 1; LoopC <= loopTo; LoopC++)
                {
                    int argEmptySpaces1 = 1024;
                    ln = ES.GetVar(npcfile, "NPC" + withBlock.Numero, "Obj" + LoopC, EmptySpaces: ref argEmptySpaces1);
                    withBlock.Invent.Object_Renamed[LoopC].ObjIndex = Convert.ToInt16(Migration.ParseVal(General.ReadField(1, ref ln, 45)));
                    withBlock.Invent.Object_Renamed[LoopC].Amount = Convert.ToInt16(Migration.ParseVal(General.ReadField(2, ref ln, 45)));

                }
            }
        }


        public static void TirarOroNpc(int Cantidad, ref Declaraciones.WorldPos Pos)
        {
            // ***************************************************
            // Autor: ZaMa
            // Last Modification: 13/02/2010
            // ***************************************************
            try
            {

                byte i;
                Declaraciones.Obj MiObj;
                int RemainingGold;
                if (Cantidad > 0)
                {

                    RemainingGold = Cantidad;

                    while (RemainingGold > 0)
                    {

                        // Tira pilon de 10k
                        if (RemainingGold > Declaraciones.MAX_INVENTORY_OBJS)
                        {
                            MiObj.Amount = Declaraciones.MAX_INVENTORY_OBJS;
                            RemainingGold = RemainingGold - Declaraciones.MAX_INVENTORY_OBJS;
                        }

                        // Tira lo que quede
                        else
                        {
                            MiObj.Amount = Convert.ToInt16(RemainingGold);
                            RemainingGold = 0;
                        }

                        MiObj.ObjIndex = Declaraciones.iORO;

                        bool argNotPirata = true;
                        TirarItemAlPiso(ref Pos, ref MiObj, NotPirata: ref argNotPirata);
                    }
                }
            }


            catch (Exception ex)
            {
                Console.WriteLine("Error in NPC_TIRAR_ITEMS: " + ex.Message);
                string argdesc = "Error en TirarOro. Error " + ex.GetType().Name + " : " + ex.Message;
                General.LogError(ref argdesc);
            }
        }
    }
}